Change snapshot url returned to UUID, migrate older snapshots

mayankkapoor opened the pull request:
Now the url returned after creating a snapshot shows the UUID instead of ID.

mayankkapoor committed:
Change snapshot url returned to UUID

mayankkapoor committed:
Change snapshot url returned to UUID

sorenh commented:
Have you thought about migration paths? If there are existing snapshots, they'll be using the id on the filesystem, so that'll need to change.

mayankkapoor commented:
Based on a conversation with @sorenh, we'll need to consider migration path for changing to uuid for filesystem paths as well. People who used the V1 API should be able to access snapshots via ID, and people who use newer APIs should be able to access via UUIDs. The recommended approach: 
Add a boolean to some of the models saying whether to expose them in the v1 api. Having a boolean that tracks if a given resource instance should work in the v1 API would help us make that decision. So an old snapshot will have 'available_in_v1_api' Boolean set to true. The newer snapshots won't.

anshprat commented:
Presently the baggage of old API is very less. As of now, is there any real user/data who is using the older API and needs to be migrated to the new API? Anyone who would crib and complain if they can no longer access old API?

sorenh commented:
It's a matter of principle and practice. It may cost us an hour or two of effort now, but we'll gain valuable insight into the potential difficulties of dealing with such issues. This will benefit us in the long run.

sorenh commented:
As discussed elsewhere, the plan is:

1) Add the visible_in_v1_api field to the snapshot model. Make it default to False, but set it to True for all current snapshots. (The easiest is probably to set default=True, run makemigrations, change it to False, run makemigrations again).
2) Make the v1 ViewSet's get_queryset filter out all snapshots that have visible_in_v1_api = False.
3) Rename all the directories on disk from <id> format to <uuid> format.
4) For all resources that have visible_in_v1_api True, provide a symlink from <id> to <uuid>.

Step 3 and 4 will be done using Django's migration framework. It's meant for database migrations, but we see no reason why we couldn't use it for this purpose as well (using RunPython migrations).

mayankkapoor committed:
Added migration path

- Add new model field ‘visible_to_v1_api’ to denote resources that are
visible to v1 API.
- To migrate old snapshots, we first rename the directory, set
visible_to_v1_api to True and create a symlink to previous directory
name so they’re accessible via V1 API.

mayankkapoor committed:
Added migration path

- Add new model field ‘visible_to_v1_api’ to denote resources that are
visible to v1 API.
- To migrate old snapshots, we first rename the directory, set
visible_to_v1_api to True and create a symlink to previous directory
name so they’re accessible via V1 API.

mayankkapoor committed:
Fixed minor issue with settings

added settings reference

mayankkapoor committed:
Fixed minor issue with settings

added settings reference

mayankkapoor committed:
Fix flake8 issues

mayankkapoor committed:
Fix flake8 issues

mayankkapoor committed:
exclude newer snapshots from v1 queryset

Now v1 ViewSet's get_queryset filters out all snapshots that have
visible_in_v1_api = False.

mayankkapoor committed:
exclude newer snapshots from v1 queryset

Now v1 ViewSet's get_queryset filters out all snapshots that have
visible_in_v1_api = False.

sorenh commented:
When you add a new field, all existing rows in the database get the default value assigned to them. That should be True (principle of least surprise: We don't want all snapshots to suddenly vanish from the v1 API (and their ID based URL's should also keep working)). However, going forward, we want the default to be False, except if the snapshot is created through the v1 API. Changing the default in a second migration doesn't change the existing data in the database. It just says what will be the default going forward.

I don't know for sure, but I have a hunch that the pypy thing is because of a leaked connection to MySQL:

In "regular" Python (CPython), file pointers are garbage collected once their last reference disappears. Like, say, if you have a method that opens a file, does some stuff, and then returns, CPython will close the file (because there are no references to the file left). With pypy, it works differently. If you don't explicitly close the file descriptor, it'll stay open until pypy happens to run its garbage collector and finds the unreferenced file objects and destroys it. That's why I always use:

   with open() as fp:
      do_stuff_with_fp
   do_stuff_that_doesn't_need_the_fp_anymore

When that "with" block terminates, the fp is closed. The fp object might not get garbage collected right then, but at least the file descriptor is closed.

I'm guessing this is the same problem, except with a connection to MySQL. I don't know, though. For now, I'll be moving pypy to "allow_failures" on Travis. I want to make sure it works (I want to use pypy eventually), but I also want to keep being able to trust the output from Travis.

mayankkapoor committed:
Break up into two migrations

visible_to_v1_api is set to True initially, then changed to False. All
old snapshots will have the field as True.

mayankkapoor committed:
Break up into two migrations

visible_to_v1_api is set to True initially, then changed to False. All
old snapshots will have the field as True.

mayankkapoor commented:
@sorenh ready for review. Made the changes suggested.

mayankkapoor committed:
Merge branch 'master' into change-snapshot-base-path-from-id-to-uuid

mayankkapoor committed:
Merge branch 'master' into change-snapshot-base-path-from-id-to-uuid

sorenisanerd commented on aasemble/django/apps/api/v1/serializers.py:
```diff
@@ -109,10 +109,14 @@ def create(self, validated_data):
                 if 'tags' in validated_data:
                     tags_data = validated_data.pop('tags')
                     snapshot = mirrorsvc_models.Snapshot.objects.create(**validated_data)
```
You could change the create call to: `create(visible_in_v1=(selff.view_prefix == 'v1'), **validated_data)`

sorenisanerd commented on aasemble/django/apps/api/v1/views.py:
```diff
@@ -108,7 +108,7 @@ def get_queryset(self):
                 else:
-                    qs = self.queryset.filter(mirrorset__owner_id=self.request.user.id)
+                    qs = self.queryset.filter(mirrorset__owner_id=self.request.user.id).exclude(visible_to_v1_api=False)
 
```
Instead, just add a new conditional:

```
if selff.view_prefix = 'v1':
    qs = qs.exclude(visible_to_v1_api=False)
```

sorenisanerd commented on aasemble/django/apps/api/v2/views.py:
```diff
@@ -29,3 +32,32 @@ def BuildViewSetFactory(selff):
+                    raise ValidationError({'detail': 'Method "PATCH" not allowed.'})
+                serializer.save(owner=self.request.user)
+
+        return SnapshotViewSet
```
If you make the changes I suggested, you don't need any of this.

sorenh commented:
We need tests to verify that snapshots created through v1 get `visible_in_v1_api=True` and snapshots created through any other version has it set to False. We also need to verify that they are correctly filtered.

mayankkapoor committed:
Optimizing code

Based on feedback to PR #135

mayankkapoor committed:
Optimizing code

Based on feedback to PR #135

mayankkapoor committed:
Add test to check visible_to_v1_api flag set correctly

Now unit tests check if the visible_to_v1_api flag is set properly.

mayankkapoor committed:
Add test to check visible_to_v1_api flag set correctly

Now unit tests check if the visible_to_v1_api flag is set properly.

mayankkapoor committed:
Add test_only_visible_snapshots_are_returned

Now test_only_visible_snapshots_are_returned checks if the correct
snapshots are returned for each API.

mayankkapoor committed:
Add test_only_visible_snapshots_are_returned

Now test_only_visible_snapshots_are_returned checks if the correct
snapshots are returned for each API.

mayankkapoor commented:
@sorenh incorporated feedback, created tests. One TODO remains for test_only_visible_snapshots_are_returned, which I'll address later. This can be merged without this imo.

mayankkapoor committed:
Fixed OrderedDict issue with tests

Now using OrderedDict to define data variable to compare response.data.

mayankkapoor committed:
Fixed OrderedDict issue with tests

Now using OrderedDict to define data variable to compare response.data.

mayankkapoor committed:
Fix API v1 test issue with timestamps

Now complete data isn’t compared in response.

mayankkapoor committed:
Fix API v1 test issue with timestamps

Now complete data isn’t compared in response.

sorenisanerd committed:
Generate contributor list

anand0707 committed:
Added test for logout

anand0707 committed:
Fixed method calling issues

sorenisanerd commented on aasemble/django/apps/api/v1/serializers.py:
```diff
@@ -125,7 +125,7 @@ def update(self, instance, validated_data):
             class Meta:
                 model = mirrorsvc_models.Snapshot
-                fields = ('self', 'timestamp', 'mirrorset')
+                fields = ('self', 'timestamp', 'mirrorset', 'visible_to_v1_api')
```
Sorry, I must have missed this when I reviewed it earlier. Why should this be part of the public API? I was thinking of it as an internal bit of book keeping.

sorenh commented:
I'm sorry I'm dragging my feet on this one. I guess I'm a bit concerned, since it makes changes to the filesystem which we haven't done before. Let's get the last wrinkled worked out, and get it merged, shall we?

sorenisanerd committed:
Merge pull request #161 from anand0707/functional-test-python--aasemble-django-logout-button-test

Functional test python  aasemble django logout button test

sorenisanerd committed:
Merge pull request #159 from sorenh/contributors

Generate contributor list

sorenisanerd committed:
Merge pull request #158 from piyush110786/test

Test case consistency implied and 1 error removed

mayankkapoor committed:
Update fixture timestamps

Now timestamps all have 000 for milliseconds.

mayankkapoor committed:
Update tests, remove visible_to_v1_api flag

Updated tests and removed visible_to_v1_api flag from being shown in API

mayankkapoor committed:
resolve merge conflicts

mayankkapoor commented:
Changes:
- removed visible_to_v1_flag from being seen in API
- Updated fixture timestamps, removed milliseconds for v1 snapshots
- Updated tests

mayankkapoor committed:
Fix OrderedDict issue

mayankkapoor committed:
Second times the charm

Fix issues with snapshot tests

mayankkapoor commented:
```
FAIL: test_snapshot_create_sets_visible_flag_properly (aasemble.django.apps.api.tests.APIv1Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/aaSemble/python-aasemble.django/aasemble/django/apps/api/tests.py", line 977, in test_snapshot_create_sets_visible_flag_properly
    self.assertEqual(snapshot_object.visible_to_v1_api, True)
AssertionError: False != True
```
@sorenh Need some advice. I'm observing some different behavior on MySQL (Travis) as opposed to SQLite. This test passes on SQLite tests. Somehow on Travis, if a new snapshot gets created using V1 API, it still gets the visible_to_v1_api flag set to False. We explicity set it to true in v1/serializers.py>SnapshotSerializer>create
```
snapshot = mirrorsvc_models.Snapshot.objects.create(visible_to_v1_api=(selff.view_prefix == 'v1'), **validated_data)
```

sorenh commented:
I think the problem is something other than you think.

You create the snapshot, then attempt to get its ID by using obj['self'][-2:-1]. If e.g. obj['self'] = `/something/63/`, then you'll get id = 3, not 63.

I'm not sure about the details, but with sqlite, I see the assigned ID be 53. With MySQL, it's 68. At least it was for me. Your mileage may vary. In either case, you're not getting the snapshot you just created.

You can either something like obj['self'].split('/')[-2] to get the whole ID.

mayankkapoor commented:
Of course. I just thought of it as well. Sorry for the bother.

mayankkapoor committed:
Fix logic bug in tests

sorenh commented:
Np. Happens to the best of us.

sorenh commented:
@mayankkapoor I'm happy to report that the migration seems to have worked as expected. Great work!

mayankkapoor commented:
@sorenh Awesome. Thanks for all the guidance.
